File: src/training-data.cc
#include "training-data.h"
#include <nan.h>
#include "fann-includes.h"
#include <iostream>
#include "utils.h"
#include "training-data.h"

namespace fanny {

class TDIOWorker : public Nan::AsyncWorker {
public:
	TrainingData *trainingData;
	std::string filename;
	bool isSave;
	bool isFixed;
	unsigned int decimalPoint;


	TDIOWorker(
		Nan::Callback *callback,
		v8::Local<v8::Object> tdHolder,
		std::string &_filename,
		bool _isSave,
		bool _isFixed,
		unsigned int _decimalPoint
	) : Nan::AsyncWorker(callback), filename(_filename),
		isSave(_isSave), isFixed(_isFixed),
		decimalPoint(_decimalPoint) {
		SaveToPersistent("tdHolder", tdHolder);
		trainingData = Nan::ObjectWrap::Unwrap<TrainingData>(tdHolder);
	}

	~TDIOWorker() {}

	void Execute() {
		if (!isSave) {
			if (!trainingData->trainingData->read_train_from_file(filename)) {
				SetErrorMessage("Error reading training data file");
			}
		} else if (!isFixed) {
			if (!trainingData->trainingData->save_train(filename)) {
				SetErrorMessage("Error saving training data file");
			}
		} else {
			if (!trainingData->trainingData->save_train_to_fixed(filename, decimalPoint)) {
				SetErrorMessage("Error saving training data file");
			}
		}
	}
};


Nan::Persistent<v8::FunctionTemplate> TrainingData::constructorFunctionTpl;

void TrainingData::Init(v8::Local<v8::Object> target) {
	// Create new function template for this JS class constructor
	v8::Local<v8::FunctionTemplate> tpl = Nan::New<v8::FunctionTemplate>(New);
	// Set the class name
	tpl->SetClassName(Nan::New("TrainingData").ToLocalChecked());
	// Set the number of "slots" to allocate for fields on this class, not including prototype methods
	tpl->InstanceTemplate()->SetInternalFieldCount(1);
	// Save a reference to the function template
	TrainingData::constructorFunctionTpl.Reset(tpl);

	// Add prototype methods
	Nan::SetPrototypeMethod(tpl, "readTrainFromFile", readTrainFromFile);
	Nan::SetPrototypeMethod(tpl, "saveTrain", saveTrain);
	Nan::SetPrototypeMethod(tpl, "saveTrainToFixed", saveTrainToFixed);
	Nan::SetPrototypeMethod(tpl, "scaleInputTrainData", scaleInputTrainData);
	Nan::SetPrototypeMethod(tpl, "scaleOutputTrainData", scaleOutputTrainData);
	Nan::SetPrototypeMethod(tpl, "scaleTrainData", scaleTrainData);
	Nan::SetPrototypeMethod(tpl, "subsetTrainData", subsetTrainData);
	Nan::SetPrototypeMethod(tpl, "shuffle", shuffle);
	Nan::SetPrototypeMethod(tpl, "merge", merge);
	Nan::SetPrototypeMethod(tpl, "length", length);
	Nan::SetPrototypeMethod(tpl, "numInput", numInput);
	Nan::SetPrototypeMethod(tpl, "numOutput", numOutput);
	Nan::SetPrototypeMethod(tpl, "getInput", getInput);
	Nan::SetPrototypeMethod(tpl, "getOutput", getOutput);
	Nan::SetPrototypeMethod(tpl, "getTrainInput", getTrainInput);
	Nan::SetPrototypeMethod(tpl, "getTrainOutput", getTrainOutput);
	Nan::SetPrototypeMethod(tpl, "setTrainData", setTrainData);
	Nan::SetPrototypeMethod(tpl, "getMinInput", getMinInput);
	Nan::SetPrototypeMethod(tpl, "getMaxInput", getMaxInput);
	Nan::SetPrototypeMethod(tpl, "getMinOutput", getMinOutput);
	Nan::SetPrototypeMethod(tpl, "getMaxOutput", getMaxOutput);

	// Assign a property called 'TrainingData' to module.exports, pointing to our constructor
	Nan::Set(target, Nan::New("TrainingData").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
}

TrainingData::TrainingData(FANN::training_data *_training_data) : trainingData(_training_data) {}

TrainingData::~TrainingData() {
	delete trainingData;
}

NAN_METHOD(TrainingData::New) {
	FANN::training_data *trainingData;
	if (info.Length() == 1 && Nan::New(TrainingData::constructorFunctionTpl)->HasInstance(info[0])) {
		TrainingData *other = Nan::ObjectWrap::Unwrap<TrainingData>(info[0].As<v8::Object>());
		FANN::training_data *otherTrainingData = other->trainingData;
		if (otherTrainingData->length_train_data() > 0) {
			trainingData = new FANN::training_data(*otherTrainingData);
		} else {
			trainingData = new FANN::training_data();
		}
	} else {
		trainingData = new FANN::training_data();
	}
	TrainingData *obj = new TrainingData(trainingData);
	obj->Wrap(info.This());
	info.GetReturnValue().Set(info.This());
}

NAN_METHOD(TrainingData::shuffle) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->shuffle_train_data();
}

NAN_METHOD(TrainingData::merge) {
	if (info.Length() != 1) {
		return Nan::ThrowError("Requires single argument");
	}
	if (!Nan::New(TrainingData::constructorFunctionTpl)->HasInstance(info[0])) {
		return Nan::ThrowError("Must be an instance of TrainingData");
	}
	TrainingData *other = Nan::ObjectWrap::Unwrap<TrainingData>(info[0].As<v8::Object>());
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->merge_train_data(*other->trainingData);
}

NAN_METHOD(TrainingData::length) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->length_train_data());
}

NAN_METHOD(TrainingData::numInput) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->num_input_train_data());
}

NAN_METHOD(TrainingData::numOutput) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->num_output_train_data());
}

NAN_METHOD(TrainingData::getInput) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	fann_type **data = self->trainingData->get_input();
	info.GetReturnValue().Set(fannDataSetToV8Array(data, self->trainingData->length_train_data(), self->trainingData->num_input_train_data()));
}

NAN_METHOD(TrainingData::getOutput) {
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	fann_type **data = self->trainingData->get_output();
	info.GetReturnValue().Set(fannDataSetToV8Array(data, self->trainingData->length_train_data(), self->trainingData->num_output_train_data()));
}

NAN_METHOD(TrainingData::getTrainInput) {
	if (info.Length() != 1 || !info[0]->IsNumber()) {
		return Nan::ThrowError("Argument must be a number");
	}
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	unsigned int pos = info[0]->Uint32Value();
	fann_type *data = self->trainingData->get_train_input(pos);
	info.GetReturnValue().Set(fannDataToV8Array(data, self->trainingData->num_input_train_data()));
}

NAN_METHOD(TrainingData::getTrainOutput) {
	if (info.Length() != 1 || !info[0]->IsNumber()) {
		return Nan::ThrowError("Argument must be a number");
	}
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	unsigned int pos = info[0]->Uint32Value();
	fann_type *data = self->trainingData->get_train_output(pos);
	info.GetReturnValue().Set(fannDataToV8Array(data, self->trainingData->num_output_train_data()));
}

NAN_METHOD(TrainingData::setTrainData) {
	if (info.Length() != 2) return Nan::ThrowError("Must have 2 arguments: input, output");
	if (!info[0]->IsArray() || !info[1]->IsArray()) return Nan::ThrowError("Not an array");
	v8::Local<v8::Array> inputs = info[0].As<v8::Array>();
	v8::Local<v8::Array> outputs = info[1].As<v8::Array>();
	unsigned int dataSetLength = inputs->Length();
	if (outputs->Length() != dataSetLength) return Nan::ThrowError("Input and output dataset sizes must match");
	if (!dataSetLength) return Nan::ThrowError("Dataset must be nonzero in size");
	std::vector<fann_type> inputVector, outputVector;
	unsigned int numInputNodes = 0, numOutputNodes = 0;
	for (unsigned int idx = 0; idx < dataSetLength; ++idx) {
		Nan::MaybeLocal<v8::Value> inputMaybeValue = Nan::Get(inputs, idx);
		Nan::MaybeLocal<v8::Value> outputMaybeValue = Nan::Get(outputs, idx);
		if (inputMaybeValue.IsEmpty() || outputMaybeValue.IsEmpty()) return Nan::ThrowError("Invalid data");
		v8::Local<v8::Value> inputValue = inputMaybeValue.ToLocalChecked();
		v8::Local<v8::Value> outputValue = outputMaybeValue.ToLocalChecked();
		if (!inputValue->IsArray() || !outputValue->IsArray()) return Nan::ThrowError("Invalid data");
		v8::Local<v8::Array> inputArray = inputValue.As<v8::Array>();
		v8::Local<v8::Array> outputArray = outputValue.As<v8::Array>();
		if (idx == 0) {
			numInputNodes = inputArray->Length();
			numOutputNodes = outputArray->Length();
			if (!numInputNodes || !numOutputNodes) return Nan::ThrowError("Invalid data");
			inputVector.reserve(dataSetLength * numInputNodes);
			outputVector.reserve(dataSetLength * numOutputNodes);
		}
		std::vector<fann_type> inputRow = v8ArrayToFannData(inputArray);
		std::vector<fann_type> outputRow = v8ArrayToFannData(outputArray);
		if (inputRow.size() != numInputNodes || outputRow.size() != numOutputNodes) return Nan::ThrowError("Invalid data");
		memcpy(&inputVector[idx * numInputNodes], &inputRow[0], numInputNodes * sizeof(fann_type));
		memcpy(&outputVector[idx * numOutputNodes], &outputRow[0], numOutputNodes * sizeof(fann_type));
	}
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->set_train_data(dataSetLength, numInputNodes, &inputVector[0], numOutputNodes, &outputVector[0]);
}

NAN_METHOD(TrainingData::getMinInput) {
	#ifndef FANNY_FIXED
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->get_min_input());
	#else
	Nan::ThrowError("Not supported for fixed fann");
	#endif
}

NAN_METHOD(TrainingData::getMaxInput) {
	#ifndef FANNY_FIXED
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->get_max_input());
	#else
	Nan::ThrowError("Not supported for fixed fann");
	#endif
}

NAN_METHOD(TrainingData::getMinOutput) {
	#ifndef FANNY_FIXED
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->get_min_output());
	#else
	Nan::ThrowError("Not supported for fixed fann");
	#endif
}

NAN_METHOD(TrainingData::getMaxOutput) {
	#ifndef FANNY_FIXED
	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	info.GetReturnValue().Set(self->trainingData->get_max_output());
	#else
	Nan::ThrowError("Not supported for fixed fann");
	#endif
}

NAN_METHOD(TrainingData::readTrainFromFile) {
	if (info.Length() < 2 || !info[0]->IsString()) return Nan::ThrowError("Filename required");
	std::string filename = std::string(*v8::String::Utf8Value(info[0]));
	Nan::Callback *callback = new Nan::Callback(info[1].As<v8::Function>());
	AsyncQueueWorker(new TDIOWorker(callback, info.Holder(), filename, false, false, 0));
}

NAN_METHOD(TrainingData::saveTrain) {
	if (info.Length() < 2 || !info[0]->IsString()) return Nan::ThrowError("Filename required");
	std::string filename = std::string(*v8::String::Utf8Value(info[0]));
	Nan::Callback *callback = new Nan::Callback(info[1].As<v8::Function>());
	AsyncQueueWorker(new TDIOWorker(callback, info.Holder(), filename, true, false, 0));
}

NAN_METHOD(TrainingData::saveTrainToFixed) {
	if (info.Length() < 3 || !info[0]->IsString() || !info[1]->IsNumber()) return Nan::ThrowError("Filename and decimalPoint required");
	std::string filename = std::string(*v8::String::Utf8Value(info[0]));
	unsigned int decimalPoint = info[1]->Uint32Value();
	Nan::Callback *callback = new Nan::Callback(info[1].As<v8::Function>());
	AsyncQueueWorker(new TDIOWorker(callback, info.Holder(), filename, true, true, decimalPoint));
}

NAN_METHOD(TrainingData::scaleInputTrainData) {
	if (info.Length() != 2) return Nan::ThrowError("Must have 2 arguments: new_min, new_max");
	if (!info[0]->IsNumber() || !info[1]->IsNumber()) return Nan::ThrowError("Arguments must be numbers");

	fann_type newMin = v8NumberToFannType(info[0]);
	fann_type newMax = v8NumberToFannType(info[1]);

	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->scale_input_train_data(newMin, newMax);
}

NAN_METHOD(TrainingData::scaleOutputTrainData) {
	if (info.Length() != 2) return Nan::ThrowError("Must have 2 arguments: new_min, new_max");
	if (!info[0]->IsNumber() || !info[1]->IsNumber()) return Nan::ThrowError("Arguments must be numbers");

	fann_type newMin = v8NumberToFannType(info[0]);
	fann_type newMax = v8NumberToFannType(info[1]);

	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->scale_output_train_data(newMin, newMax);
}

NAN_METHOD(TrainingData::scaleTrainData) {
	if (info.Length() != 2) return Nan::ThrowError("Must have 2 arguments: new_min, new_max");
	if (!info[0]->IsNumber() || !info[1]->IsNumber()) return Nan::ThrowError("Arguments must be numbers");

	fann_type newMin = v8NumberToFannType(info[0]);
	fann_type newMax = v8NumberToFannType(info[1]);

	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->scale_train_data(newMin, newMax);
}

NAN_METHOD(TrainingData::subsetTrainData) {
	if (info.Length() != 2) return Nan::ThrowError("Must have 2 arguments: pos, legth");
	if (!info[0]->IsNumber() || !info[1]->IsNumber()) return Nan::ThrowError("Arguments must be numbers");

	unsigned int pos = info[0]->Uint32Value();
	unsigned int length = info[1]->Uint32Value();

	TrainingData *self = Nan::ObjectWrap::Unwrap<TrainingData>(info.Holder());
	self->trainingData->subset_train_data(pos, length);
}

}

File: lib/training-data.js
// Copyright 2016 Zipscene, LLC
// Licensed under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

var utils = require('./utils');
var XError = require('xerror');

function TrainingData(fannyTrainingData, datatype) {
	this._fannyTrainingData = fannyTrainingData;
	this._datatype = datatype;
}

TrainingData.prototype.clone = function() {
	var addon = utils.getAddon(this._datatype);
	var fannyTrainingData = new addon.TrainingData(this._fannyTrainingData);
	return new TrainingData(fannyTrainingData, this._datatype);
};

// Possible invocations:
// setData(<ArrayOfPairs>)
// setData(<InputsArrayOfArrays>, <OutputsArrayOfArrays>)
// For ArrayOfPairs, it should be an array of objects.  Each object should
// contain the keys "input" and "output", and each should be an array of
// the appropriate size.
TrainingData.prototype.setData = function(arg1, arg2) {
	var inputs = [];
	var outputs = [];
	var i, entry;
	if (Array.isArray(arg1) && !arg2) {
		for (i = 0; i < arg1.length; ++i) {
			entry = arg1[i];
			if (Array.isArray(entry) && entry.length === 2 && Array.isArray(entry[0]) && Array.isArray(entry[1])) {
				inputs.push(entry[0]);
				outputs.push(entry[1]);
			} else if (typeof entry === 'object' && entry && Array.isArray(entry.input) && Array.isArray(entry.output)) {
				inputs.push(entry.input);
				outputs.push(entry.output);
			} else {
				throw new XError(XError.INVALID_ARGUMENT);
			}
		}
	} else if (Array.isArray(arg1) && Array.isArray(arg2) && Array.isArray(arg1[0])) {
		if (arg1.length !== arg2.length) throw new XError(XError.INVALID_ARGUMENT);
		for (i = 0; i < arg1.length; ++i) {
			if (!Array.isArray(arg1[i]) || !Array.isArray(arg2[i])) throw new XError(XError.INVALID_ARGUMENT);
			inputs.push(arg1[i]);
			outputs.push(arg2[i]);
		}
	} else {
		throw new XError(XError.INVALID_ARGUMENT);
	}
	try {
		this._fannyTrainingData.setTrainData(inputs, outputs);
	} catch (ex) {
		throw new XError(XError.INVALID_ARGUMENT, ex);
	}
};

// If fixedDecimalPoint is set, it's saved to a fixed format
TrainingData.prototype.save = function(filename, fixedDecimalPoint) {
	var self = this;
	if (typeof filename !== 'string') throw new XError(XError.INVALID_ARGUMENT);
	return new Promise(function(resolve, reject) {
		var cb = function(err) {
			if (err) return reject(new XError(err));
			resolve();
		};
		if (typeof fixedDecimalPoint === 'number') {
			self._fannyTrainingData.saveTrainToFixed(filename, fixedDecimalPoint, cb);
		} else {
			self._fannyTrainingData.saveTrain(filename, cb);
		}
	});
};

TrainingData.prototype.shuffle = function() {
	this._fannyTrainingData.shuffle();
};

TrainingData.prototype.merge = function(data) {
	if (!data || !data._fannyTrainingData) {
		throw new XError(XError.INVALID_ARGUMENT, 'data must be an instanceof TrianingData');
	}
	return this._fannyTrainingData.merge(data._fannyTrainingData);
};

TrainingData.prototype.getLength = function() {
	return this._fannyTrainingData.length();
};

TrainingData.prototype.getNumInputs = function() {
	return this._fannyTrainingData.numInput();
};

TrainingData.prototype.getNumOutputs = function() {
	return this._fannyTrainingData.numOutput();
};

TrainingData.prototype.getInputData = function() {
	return this._fannyTrainingData.getInput();
};

TrainingData.prototype.getOutputData = function() {
	return this._fannyTrainingData.getOutput();
};

TrainingData.prototype.getOneInputData = function(pos) {
	if (typeof pos !== 'number') {
		throw new XError(XError.INVALID_ARGUMENT, 'argument should be a number');
	}
	return this._fannyTrainingData.getTrainInput(pos);
};

TrainingData.prototype.getOneOutputData = function(pos) {
	if (typeof pos !== 'number') {
		throw new XError(XError.INVALID_ARGUMENT, 'argument should be a number');
	}
	return this._fannyTrainingData.getTrainOutput(pos);
};

TrainingData.prototype.getMinInput = function() {
	return this._fannyTrainingData.getMinInput();
};

TrainingData.prototype.getMaxInput = function() {
	return this._fannyTrainingData.getMaxInput();
};

TrainingData.prototype.getMinOutput = function() {
	return this._fannyTrainingData.getMinOutput();
};

TrainingData.prototype.getMaxOutput = function() {
	return this._fannyTrainingData.getMaxOutput();
};

TrainingData.prototype.scaleInput = function(min, max) {
	if (typeof min !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'min must be a number');
	if (typeof max !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'max must be a number');
	return this._fannyTrainingData.scaleInputTrainData(min, max);
};

TrainingData.prototype.scaleOutput = function(min, max) {
	if (typeof min !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'min must be a number');
	if (typeof max !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'max must be a number');
	return this._fannyTrainingData.scaleOutputTrainData(min, max);
};

TrainingData.prototype.scale = function(min, max) {
	if (typeof min !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'min must be a number');
	if (typeof max !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'max must be a number');
	return this._fannyTrainingData.scaleTrainData(min, max);
};

TrainingData.prototype.subset = function(startPos, length) {
	if (typeof startPos !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'startPos must be a number');
	if (typeof length !== 'number') throw new XError(XError.INVALID_ARGUMENT, 'length must be a number');
	return this._fannyTrainingData.subsetTrainData(startPos, length);
};

function createTrainingData(arg1, arg2, datatype) {
	if (typeof arg2 === 'string') {
		datatype = arg2;
		arg2 = undefined;
	}
	if (!datatype) datatype = 'float';
	var addon = utils.getAddon(datatype);
	var fannyTrainingData = new addon.TrainingData();
	var td = new TrainingData(fannyTrainingData, datatype);
	td.setData(arg1, arg2);
	return td;
}

function loadTrainingData(filename, datatype) {
	if (!filename) throw new XError(XError.INVALID_ARGUMENT, 'filename is required');
	if (!datatype) datatype = 'float';
	var addon = utils.getAddon(datatype);
	return new Promise(function(resolve, reject) {
		var fannyTrainingData = new addon.TrainingData();
		fannyTrainingData.readTrainFromFile(filename, function(err) {
			if (err) return reject(new XError(err));
			var td = new TrainingData(fannyTrainingData, datatype);
			resolve(td);
		});
	});
}

module.exports = {
	createTrainingData: createTrainingData,
	loadTrainingData: loadTrainingData
};

